language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs

```diff
@@ -24,27 +25,32 @@ pub const POP_EMPTY_VEC: u64 = NFE_VECTOR_ERROR_BASE + 2;
 pub const DESTROY_NON_EMPTY_VEC: u64 = NFE_VECTOR_ERROR_BASE + 3;
 
 #[allow(dead_code)]
-fn get_mut_vector(v: &mut NativeStructValue) -> Option<&mut NativeVector> {
+fn get_mut_vector(v: &mut NativeStructValue) -> VMResult<&mut NativeVector> {
     match v {
-        NativeStructValue::Vector(v) => Some(v),
+        NativeStructValue::Vector(v) => Ok(v),
     }
 }
 
-fn get_vector(v: &NativeStructValue) -> Option<&NativeVector> {
+fn get_vector(v: &NativeStructValue) -> VMResult<&NativeVector> {
     match v {
-        NativeStructValue::Vector(v) => Some(v),
+        NativeStructValue::Vector(v) => Ok(v),
     }
 }
 
 macro_rules! get_vector_ref {
     ($args: expr) => {
-        match $args
-            .pop_front()
-            .and_then(|v| v.value_as::<ReferenceValue>())
-        {
-            Some(v) => v,
-            None => return NativeReturnStatus::InvalidArguments,
-        }
+        match $args.pop_front() {
+            Some(v) => match v.value_as::<ReferenceValue>() {
+                Ok(reference) => reference,
+                Err(err) => return NativeReturnStatus::InvariantError(err),
+            },
+            None => {
+                return NativeReturnStatus::InvariantError(
+                    VMStatus::new(StatusCode::UNREACHABLE)
```

Comment:
Should it be unreachable or argument length mismatch or sth?

Reply:
good point. The reason I have unreachable on a lot of those is because there is already a check for argument count in the common dispatch in the interpreter so this should really be unreachable. 
Overall I want to review the native dispatching a bit and maybe this is a place to introduce some Mirai assume or something, not too sure at the moment....

Reply:
Had some offline conversations. The reason why this is `UNREACHABLE` is because we've already checked the argument length at the beginning of the dispatcher so it makes sense to me.